log4j-api/src/main/java/org/apache/logging/log4j/spi/AbstractLogger.java

```diff
@@ -2762,12 +2746,8 @@ public LogBuilder atLevel(final Level level) {
      * @since 2.20.0
      */
     protected LogBuilder getLogBuilder(Level level) {
-        if (Constants.ENABLE_THREADLOCALS) {
-            DefaultLogBuilder builder = logBuilder.get();
-            if (!builder.isInUse()) {
-                return builder.reset(this, level);
-            }
-        }
-        return new DefaultLogBuilder(this, level);
+        DefaultLogBuilder builder = recycler.acquire();
```

Comment:
Where is the log builder released?

Reply:
I thought about introducing a `RecyclerAware` interface with a `setRecycler` method. Classes like the `DefaultLogBuilder` or `ReusableParameterizedMessage` would implement it.

Of course your solution also works.

Reply:
I wrote the `RecyclerAware` idea here: https://github.com/ppkarwasz/logging-log4j2/tree/pkarwasz/recycler-api-3.x

Reply:
I have my reservations for making sure each recycler factory specializes on instances of type `RecyclerAware`, though I couldn't resist the utility of this idea. :star_struck: